components/mediation/registry/org.wso2.micro.integrator.registry/src/main/java/org/wso2/micro/integrator/registry/MicroIntegratorRegistryListener.java

```diff
@@ -0,0 +1,159 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wso2.micro.integrator.registry;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.file.FileSystems;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.HashMap;
+import java.util.Map;
+
+public class MicroIntegratorRegistryListener implements Runnable {
+
+    private static final Log log = LogFactory.getLog(MicroIntegratorRegistryListener.class);
+    private final WatchService watcher;
+    private final Map<WatchKey, Path> keys;
+
+    /**
+     * Creates a WatchService and registers the given directory
+     */
+    public MicroIntegratorRegistryListener(Path dir) throws IOException {
+        this.watcher = FileSystems.getDefault().newWatchService();
+        this.keys = new HashMap<>();
+
+        walkAndRegisterDirectories(dir);
+    }
+
+    /**
+     * Register the given directory with the WatchService; This function will be called by FileVisitor
+     *
+     * @param dir registry file location
+     */
+    private void registerDirectory(Path dir) throws IOException
+    {
+        WatchKey key = dir.register(watcher,
+                StandardWatchEventKinds.ENTRY_CREATE,
+                StandardWatchEventKinds.ENTRY_DELETE,
+                StandardWatchEventKinds.ENTRY_MODIFY);
+        keys.put(key, dir);
+    }
+
+
+    /**
+     * Register the given directory, and all its sub-directories, with the WatchService.
+     *
+     * @param start the root path
+     */
+    private void walkAndRegisterDirectories(final Path start) throws IOException {
+        // register directory and sub-directories
+        Files.walkFileTree(start, new SimpleFileVisitor<Path>() {
+            @Override
+            public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+                registerDirectory(dir);
+                return FileVisitResult.CONTINUE;
+            }
+        });
+    }
+
+    public void run() {
+        try {
+            WatchKey key;
+            while ((key = watcher.take()) != null) {
+                Path dir = keys.get(key);
+                if (dir == null) {
+                    log.error("WatchKey for listening to registry resources changes not recognized!!");
```

Comment:
Do we need to do something like invalidating the key or something? Basically will the listener recover if it comes to this state?

Reply:
Yes, the task is continued after that.. So the listener will take another key when an event happen.